### ISSUE TITLE
Franz's list of random TODOs and issues

### ISSUE THREAD

**Issue by dictoon** [appleseedhq/appleseed#2762]
A good resolution I'm taking for 2020 is to stop having a parallel, private list of issues and things to do, and use GitHub's issue tracker instead.

What you'll find below is a random list of TODOs and issues, some of which might be already done, obsolete, or just plain bad ideas. They are roughly sorted in ascending creation time, i.e. top issues are _very_ old and bottom ones are newer.

These entries, as they are tackled, should be converted one by one to GitHub tickets.

---

- [ ] adaptive sampling in path tracer
- [ ] missing parse error exceptions in objmeshfilelexer.h
- [ ] measure performance of compute_motion_bboxes() for large trees
- [ ] add rays/second stats
- [ ] update docs (parameters)
- [ ] hsl / hsv: h in [0,1] instead of [0,360) -> see standard
- [ ] indirect range (limit gi to short range)
- [ ] MIS bug when max_path_length is set to 1 in PT? -> retest
- [ ] allow choosing EXR compression mode in appleseed and in tools
- [ ] print warning if hdr texture not in linear RGB?
- [ ] can't stop progressive rendering during first pass
- [ ] update acc. structures for each frame, not just when reinitializing rendering
- [ ] make size of triangle tree cache configurable
- [ ] support ordering statistics
- [ ] wally: change default color space for textures to linear RGB? -> make it a default setting, same for colors
- [ ] support transformation motion of lights themselves
- [ ] on Linux, set CMAKE_SKIP_RPATH and CMAKE_SKIP_INSTALL_RPATH to true in Ship configuration
- [ ] read-only settings window to allow peeking at settings during rendering?
- [ ] plastic teapot: checkerboard sampling artifacts?
- [ ] plastic teapot: slow! with DRT, 64 spp, 0 bounces: 50 seconds!
- [ ] modify appleseed.package to make .tar.gz archives on linux/osx
- [ ] mescaline_10_14: frame 118 fucked up because geo at frame 119 is undefined?
- [ ] objmeshfilewriter should take per-face materials into account
- [ ] add test scenes for BinaryMesh format; alpha mapping without intersection filters
- [ ] updateprojectfile should have an option to remove unused entities
- [ ] no-shadow diagnostic mode?
- [ ] colored fresnel
- [ ] add support for integer pixel formats in OpenEXR image file writer
- [ ] cache bvh on disk
- [ ] crash when picking during reconstruction of assembly bvh
- [ ] build intersection filters in parallel with the construction of the bvh (use all cores; pool in case multiple triangle trees are built at the same time)
- [ ] optimization: mark "zero" AOVs to avoid multiplying/adding/clamping them
- [ ] generalize "alpha mode" to "scalar mode"?
- [ ] fix all get_memory_size()
- [ ] add an option to compute aovs in interactive render mode
- [ ] sbvh works?
- [ ] assertion in sbvh on "tesla_coil_in_ri.0001 - sbvh.appleseed" UPDATE 2014/02/25: still crashes
- [ ] lots of optimizations possible if we could know if a BSDF is purely specular!
- [ ] add "cast indirect" flag to environment?
- [ ] allow to skip lights whose max contribution is below a user-specified threshold (http://www.luxrender.net/forum/viewtopic.php?f=8&t=10315&sid=daa8bd4b672075a39d712d5515d03210)
- [ ] zooming in/out with +/- doesn't work after picking an object (because focus is in the explorer)
- [ ] allow scheduling multiple jobs at once
- [ ] optimize rand_int1()
- [ ] try photon relaxation
- [ ] add support for moving photons (see PM at Pixar slides)
- [ ] test adaptive photon tracing based on visibility
- [ ] reorder photon data to increase cache locality?
- [ ] when sampling bsdf_mix and bsdf_blend, evaluate the non-sampled bsdf if the sampled bsdf is not purely specular? -> why not, suggested by elvic but careful about caustics when they're disabled! see issue on github
- [ ] interactive sppm? need to regularly build a new photon map
- [ ] add ShadingPoint::get_original_shading_basis()
- [ ] material with edf+bsdf = fireflies sometimes?
- [ ] implement Light Near Start and Light Low Threshold (http://www.luxrender.net/forum/viewtopic.php?f=34&t=10529) [issue]
- [ ] new attempt at prefetching? http://www.naftaliharris.com/blog/2x-speedup-with-one-line-of-code/
- [ ] allow to disable lights/edfs easily -> #581
- [ ] studio: crash after picking and modifying object instance? confirmed 2014-12-23
- [ ] "final average luminance is 1.#INF00" when using straight alpha? not due to straight alpha...
- [ ] add a flag to edfs to force them to be sampled
- [ ] picking should clear the entity filter [issue]
- [ ] sampling a mixture of mesh lights where some have many polygons and some have few results in massive noise for those that have few polygons -> importance sampling should solve that, was it tried?
- [ ] python scripts: allow to set the path to the appleseed tool via an env variable -> or rather, path to a bin directory since we need many tools
- [ ] python scripts: move common code to separate module
- [ ] difference brightness with dl_light_samples set to 1 vs 4?
- [ ] prefix all preprocessor symbols with APPLESEED_ (at least those defined in CMakeLists.txt, such as USE_QMC_SAMPLER or ARCH32)
- [ ] Quaternion::extract_axis_angle() -> extract_rotation() ?
- [ ] document limitation: can only instantiate objects of current assembly, not of parent assemblies
- [ ] bug: light near start not honored by ptne when hitting emitter?
- [ ] in studio: add configuration combobox to toolbar; remove interactive render (F6) and render using the selected config instead. also configure the shading override menu based on selected config. -> #580
- [ ] quicksaves get overwritten when main output is absent
- [ ] set default value of ParamArray parameters to an empty ParamArray
- [ ] sse1 bvh traversal -> #578
- [ ] profile curve rendering
- [ ] use geometry hash instead of assembly version ids in assemblytree.cpp?
- [ ] use raytracing for IBL when using Ray Traced direct lighting mode in SPPM -> #536
- [ ] merge sppmlightingengine - Copy.cpp
- [ ] foundation::Tile::set_component() API is confusing and dangerous!
- [ ] remove redundancy between MeshObject and StaticTessellation -> #574
- [ ] in .studio, text boxes are always scrolled to the right
- [ ] investigate optimization opportunities in replacing std::sort() by more efficient algorithms for the actual data sets we're concerned with
- [ ] trim trailing blank lines in SeExpr expressions?
- [ ] implement auto-framing (ctrl+shift+f)
- [ ] automatically adjust the film's aspect ratio when the resolution is changed?
- [ ] allow to choose another bvh partitioner (e.g. a plain median split one) for faster bvh builds
- [ ] add case sensitivity mode next to regex filters in .studio UI
- [ ] uniformize entity human readable names
- [ ] can't reload project file with accentuated characters
- [ ] when encountering singular matrix during parsing, print name of entity being parsed
- [ ] object instance item *in focus* should be pink when object instance has no materials
- [ ] allow importing multiple Disney materials at once
- [ ] implement Select All (for entities of a given type)
- [ ] allow assigning materials to all slots of an object instance at once
- [ ] allow saving a project during rendering?
- [ ] immediately commit changes when folding/unfolding Disney material layers
- [ ] immediately commit changes when renaming Disney material layers
- [ ] panels in attribute/entity editors should be draggable like in 3ds Max
- [ ] add a quick way to enable/disable layers in Disney material, without having to change mask
- [ ] add a way to duplicate an entity (e.g. a Disney material)
- [ ] make convertmeshfile incremental (object per object), should be easy
- [ ] compare our implementation of Bezier curve intersection with pbrt's (https://github.com/mmp/pbrt-v3/blob/master/src/shapes/curve.cpp)
- [ ] improve scheduling of tiles in multipass rendering
- [ ] define unary + for Vectors, to be able to write w = backside ? -n : +n;
- [ ] speed up microfacets unit tests (3.5 seconds in release)
- [ ] importing twice the same texture, then deleting one of them results in crash of .studio
- [ ] pivot point set with 'f' not saved to project? also does not mark project as dirty
- [ ] proper handling of multiscreens (remember on which screen .studio and each of its window was last open)
- [ ] implement consistent normal interpolation: http://blog.yiningkarlli.com/2015/01/consistent-normal-interpolation.html
- [ ] investigate precision and performance gains of using _mm_rsqrt_ss() for inverse square root (possibly followed by one or more Newton-Raphson iterations)
- [ ] ctrl+shift+0 no longer works? (#1042, #2759)
- [ ] pressing Delete twice rapidly on e.g. an object instance item leads to crash (two events triggered, 2nd one on a no-longer-existing item)
- [ ] show/hide on assembly instances
- [ ] sss: add reduced intensity term S(0)
- [ ] sss: add single scattering term S(1)
- [ ] sss: fix black outlines in SSS; see Peter Kutz blog post
- [ ] sss: in cube test scene, SSS's next event estimation requires back material of light object, while pure path tracing code requires front material? -> objs have no normals, everything is front side?
- [ ] sss: GI won't work if mesh has no vertex normals (because unable to distinguish front/back faces)
- [ ] sss: build a bvh per sss material to speed up probe ray tracing
- [ ] sss: separate Rmax radius per color channel? randomly choosing Rmax?
- [ ] sss: we need to use get_biased_point() to get ray origins, different than offset-and-refined point (but does that really work?)
- [ ] sss: add debug2 scene to test suite
- [ ] sss: move BSDF::DiracDelta to ScatteringModes
- [ ] sss: fix banding artifacts with ND
- [ ] sss: evaluate the BSSRDF only once when it is directional (already makes sense: already evaluated twice for a given shading point, in pathtracer.h and ptlightingengine.h)
- [ ] sss: make sss work with sppm
- [ ] sss: "if there are multiple intersections we build a conditional CDF based on the distance to each point, then stochastically choose a point to sample"
- [ ] improve interface of foundation::Dual3d to avoid expensive copies
- [ ] warnings & optimization flags on Linux
- [ ] OSLBSDF shouldn't be wrapped into BSDFWrapper? double wrapping?
- [ ] artifacts/blinkings when zooming/panning light paths view?
- [ ] faster updates when changing false colors settings
- [ ] export from 3ds Max: textures are not copied
- [ ] when intersecting procedural objects, initialize result.m_distance to the t value found so far
- [ ] fresnel multiplier with specular BTDF?
- [ ] specular BTDF leads to gray surfaces, while glass BSDF doesn't?
- [ ] improve logging around loading plugins (seems like some paths from APPLESEED_SEARCHPATH are silently ignored when containing accentuated chars?)
- [ ] help @karthik find out what works and what doesn't in MRI and update MRI test scenes and ref images
- [ ] test scene aovs/01 asserts in debug -> #2504
- [ ] test scene bssrdf/099 asserts in debug -> #2505
- [ ] test scene volume/03 asserts in debug -> #2506
- [ ] direct light samples decay rate
- [ ] update copyright notices for 2020
- [ ] constrain foundation::next_multiple() and similar functions to integer types
- [ ] when linking appleseed.py2.obj in Ship configuration:
-       LINK : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance
- [ ] crash with Title_Render_Appleseed_V07.appleseed scene when switching to Albedo mode with a render region
- [ ] re-number test scenes and improve consistency (naming, subfolders, etc.)
- [ ] rename RegularSpectrum to CombSpectrum?
- [ ] factorize common code between compute_smooth_vertex_tangents_base_pose() and compute_smooth_vertex_tangents_pose()
- [ ] curves\01 - single strand shadows.appleseed: evaluate_tangent() returns 0,0,z, bitangent=0,0 -> normalize fails
- [ ] use APPLESEED_SIMD4_ALIGN everywhere instead of APPLESEED_ALIGN(16)
- [ ] "// Switch to the BSSRDF's BSDF." -> update all vertex members, including m_edf etc.!
- [ ] m_cos_on fishy: should be always > 0, can be computed in get_bsdf_prob_area() where it's used
- [ ] is indirect lighting correctly disabled in PT without NEE? probably not
- [ ] sppm/04 does not match pt/ptne reference, even with lambertian brdf
- [ ] flags handling in lights
